dlrover/python/master/monitor/metric_context.py

```diff
@@ -0,0 +1,317 @@
+    Attributes:
+        gpu_free_mem(int, MB): free gpu memory
+        gpu_used_mem(int, MB): used gpu memory
+        gpu_util(int): gpu utilization
+        gpu_temperature(int): gpu temperature
+        gpu_sm_util(float, percent): gpu sm utilization
+        gpu_tensor_util(float, percent): gpu tensor utilization
+    """
+
+    def __init__(
+        self,
+        gpu_free_mem=0,
+        gpu_used_mem=0,
+        gpu_util=0,
+        gpu_temperature=0,
+        gpu_sm_util=0.0,
+        gpu_tensor_util=0.0,
+    ):
+        super().__init__("nvidia.GPU")
+        self.metrics = {
+            GpuMetricType.GPU_FREE_MEM: gpu_free_mem,
+            GpuMetricType.GPU_USED_MEM: gpu_used_mem,
+            GpuMetricType.GPU_UTIL: gpu_util,
+            GpuMetricType.GPU_TEMP: gpu_temperature,
+            GpuMetricType.GPU_SM_UTIL: gpu_sm_util,
+            GpuMetricType.GPU_TENSOR_UTIL: gpu_tensor_util,
+        }
+
+    def set_metric(self, key, value):
+        if key in self.metrics.keys():
+            self.metrics[key] = value
+
+    def get_metric(self, key):
+        if key in self.metrics.keys():
+            return self.metrics[key]
+        else:
+            return None
+
+
+class NpuMetric(XpuMetric):
+    """NPU Metric
+    Attributes:
+        npu_total_mem(int, MB): total npu memory
+        npu_used_mem(int, MB): used npu memory
+        npu_util(int): npu utilization
+        npu_temperature(int): npu temperature
+        npu_health_state(0 or 1): npu health state
+        npu_link_state(0 or 1): npu link state
+        npu_optical_state(0 or 1): npu optical state
+        npu_network_state(0 or 1): npu network state
+        npu_chip_info_bandwidth_rx(float, MB/s): RDMA rx bandwidth
+        npu_chip_info_bandwidth_tx(float, MB/s): RDMA tx bandwidth
+    """
+
+    def __init__(
+        self,
+        npu_total_mem=0,
+        npu_used_mem=0,
+        npu_util=0,
+        npu_temperature=0,
+        npu_health_state=1,
+        npu_link_state=1,
+        npu_optical_state=1,
+        npu_network_state=1,
+        npu_tx=0.0,
+        npu_rx=0.0,
+    ):
+        super().__init__("ascend.NPU")
+        self.metrics = {
+            NpuMetricType.NPU_TOTAL_MEM: npu_total_mem,
+            NpuMetricType.NPU_USED_MEM: npu_used_mem,
+            NpuMetricType.NPU_UTIL: npu_util,
+            NpuMetricType.NPU_TEMP: npu_temperature,
+            NpuMetricType.NPU_HEALTH_STATE: npu_health_state,
+            NpuMetricType.NPU_LINK_STATE: npu_link_state,
+            NpuMetricType.NPU_OPTICAL_STATE: npu_optical_state,
+            NpuMetricType.NPU_NETWORK_STATE: npu_network_state,
+            NpuMetricType.NPU_RDMA_TX: npu_tx,
+            NpuMetricType.NPU_RDMA_RX: npu_rx,
+        }
+
+    def set_metric(self, key, value):
+        if key in self.metrics.keys():
+            self.metrics[key] = value
+
+    def get_metric(self, key):
+        if key in self.metrics.keys():
+            return self.metrics[key]
```

Comment:
The metric is usually time series data and we need to record the reported time.

Reply:
please look at the JobMetricContext::_xpu_job_metrics, it is an OrderedDict with timestamp as key, following the FIFO rule. We only keep maximum 30 job metric snapshots in 30 minutes.